src/eigsolve/biarnoldi.jl

```diff
@@ -0,0 +1,324 @@
+            fact = expand!(iter, fact; verbosity=alg.verbosity)
+            V, W = basis(fact)
+
+            # update M with the new basis vectors
+            for i in 1:K
+                MM[i, K + 1] = inner(W[i], V[K + 1])
+                MM[K + 1, i] = inner(W[K + 1], V[i])
+            end
+            MM[K + 1, K + 1] = inner(W[K + 1], V[K + 1])
+
+            numops += 1
+        else # shrink
+            numiter == maxiter && break
+
+            # Determine how many to keep
+            keep = div(3 * krylovdim + 2 * converged, 5) # strictly smaller than krylovdim since converged < howmany <= krylovdim, at least equal to converged
+            while keep < krylovdim &&
+                (isapprox(valuesH[pH[keep]], conj(valuesH[pH[keep + 1]]); rtol=1e-3) ||
+                 isapprox(valuesK[pK[keep]], conj(valuesK[pK[keep + 1]]); rtol=1e-3))
+                # increase the number of eigenvalues kept such that the eigenspace is not fractured at a eigenvalue pair
+                keep += 1
+            end
+
+            # Setp 10 & 11 - Correct the kept part of H and K and the residual 
+
+            # We know that 
+            #   Vm* residual(V) = Q_1* (Vl* residual(V)) = Q_1* Vl* (residual - Vl Ml^-1 Wl* residual) = -Q_1* Ml^-1 Wl* residual = -Q_1* MWv 
+            # as Vl*Vl = Id and Vl* residual = 0
+
+            Vv = -adjoint(Q[:, 1:keep]) * MWv
+            Ww = -adjoint(Z[:, 1:keep]) * MVw
+
+            _H[1:keep, 1:keep] += Vv * transpose(_h[1:keep])
+            _K[1:keep, 1:keep] += Ww * transpose(_k[1:keep])
+
+            # newresidual = (I - Vm Vm*) oldresidual = (I - Vl Q1 Vm*) oldresidual = oldresidual + Vl Q1 Q_1^* MWv = oldresidual + Vl Q1 Vv
+            Q1Vv = Q[:, 1:keep] * Vv
+            Z1Ww = Z[:, 1:keep] * Ww
+
+            V, W = basis(fact)
+            for i in eachindex(Q1Vv)
+                rV = add!!(rV, V[i], -Q1Vv[i])
+                rW = add!!(rW, W[i], -Z1Ww[i])
+            end
+
+            βpv = norm(rV)
+            βpw = norm(rW)
+
+            _h .*= βpv
+            _k .*= βpw
+
+            # Restore Arnoldi form in the first keep columns; this is not part of the original paper
+            _restorearnoldiformandupdatebasis!(keep, _H, Q, _h, rayleighquotient(fact)[1],
+                                               V, rV, βpv)
+            _restorearnoldiformandupdatebasis!(keep, _K, Z, _k, rayleighquotient(fact)[2],
+                                               W, rW, βpw)
+
+            # Update M according to the transformation M -> Z'MQ to save some inner products later
+            _M = view(MM, 1:keep, 1:keep)
+            _temp = view(temp, 1:keep, 1:K)
+            mul!(_temp, (Z[:, 1:keep])', M)
+            mul!(_M, _temp, Q[:, 1:keep])
+
+            # Shrink Arnoldi factorization
+            fact = shrink!(fact, keep; verbosity=alg.verbosity)
+            numiter += 1
+        end
+    end
+
+    return S, Q, T, Z, fact, converged, numiter, numops
```

Comment:
How would you feel about grouping the arguments:
```suggestion
    return (S, Q), (T, Z), fact, converged, numiter, numops
```

This way, `_schursolve` has the same number of output arguments for the different algorithms. This doesn't matter too much, but it can be nice to be able to always assume that `converged` is the 4th output, for example.